scala/lambdas/ingest-event-aggregator/src/main/scala/uk/gov/nationalarchives/eventaggregator/Aggregator.scala

```diff
@@ -0,0 +1,104 @@
+
+    def writeToLockTable(input: Input, config: Config, batchId: String)(using dynamoClient: DADynamoDBClient[F]): F[Int] = {
+      dynamoClient.writeItem(
+        DADynamoDbWriteItemRequest(
+          config.lockTable,
+          Map("messageId" -> toDynamoString(input.id.toString), "batchId" -> toDynamoString(batchId), "message" -> toDynamoString(input.asJson.printWith(Printer.noSpaces)))
+        )
+      )
+    }
+
+    private def startNewBatch(ref: Ref[F, Map[String, Batch]], sourceId: String, config: Config, lambdaTimeoutTime: Long)(using
+        sfnClient: DASFNClient[F],
+        enc: Encoder[SFNArguments]
+    ): F[String] = {
+      val expires = lambdaTimeoutTime + (config.maxSecondaryBatchingWindow * 1000)
+      val delay = Math.ceil((expires - Generators[F].generateInstant.toEpochMilli).toDouble / 1000).toInt
+      val groupId = s"${config.sourceSystem}_${Generators[F].generateUuid}"
+      val batchId = s"${groupId}_0"
+      for {
+        batchCache <- ref.update(batchCache => batchCache + (sourceId -> Batch(batchId, Instant.ofEpochMilli(expires), 1)))
+        _ <- sfnClient.startExecution(config.sfnArn, SFNArguments(groupId, batchId, delay), batchId.some)
+      } yield batchId
+    }
+
+    private def getBatchId(ref: Ref[F, Map[String, Batch]], config: Config, sourceId: String, lambdaTimeoutTime: Long)(using Encoder[SFNArguments]): F[String] = {
+      def log = logWithReason(sourceId)
+      ref.get.flatMap { batchCache =>
+        val potentialCurrentBatchForSource = batchCache.get(sourceId)
+        potentialCurrentBatchForSource match
+          case None => log(NoExistingBatch) >> startNewBatch(ref, sourceId, config, lambdaTimeoutTime)
+          case Some(currentBatchForSource) =>
+            if lambdaTimeoutTime >= currentBatchForSource.expires.toEpochMilli then log(ExpiryBeforeLambdaTimeout) >> startNewBatch(ref, sourceId, config, lambdaTimeoutTime)
+            else if currentBatchForSource.items > config.maxBatchSize then log(MaxBatchSizeExceeded) >> startNewBatch(ref, sourceId, config, lambdaTimeoutTime)
+            else ref.update(_ + (sourceId -> currentBatchForSource.copy(items = currentBatchForSource.items + 1))) >> Async[F].pure(currentBatchForSource.batchId)
+      }
+    }
+
+    override def aggregate(config: Config, ref: Ref[F, Map[String, Batch]], messages: List[SQSMessage], remainingTimeInMillis: Int)(using
+        Encoder[SFNArguments],
+        Decoder[Input]
+    ): F[List[Outcome[F, Throwable, Unit]]] = {
+      val lambdaTimeoutTime = Generators[F].generateInstant.toEpochMilli + remainingTimeInMillis
+      for {
+        fibers <- messages.parTraverse { record =>
+          val process = for {
+            batchId <- getBatchId(ref, config, record.getEventSourceArn, lambdaTimeoutTime)
+            input <- Async[F].fromEither(decode[Input](record.getBody))
+            _ <- writeToLockTable(input, config, batchId)
```

Comment:
Not sure now I think about it. @kurtismash any thoughts?

Reply:
`group` please. We're already using `batch` within our generic ingest so we'll use `group` here to avoid confusion. Aggregate into groups, ingest in batches. (As DynamoDB's GSI process may cause an ingest batch to be missing some items that were part of the aggregation group on the first try).

Reply:
So just to double check, everything in this lambda should be `group` with the exception of the "TDR_$uuid_0" reference that's passed to the step function?

Reply:
> So just to double check, everything in this lambda should be group with the exception of the "TDR_$uuid_0" reference that's passed to the step function?

Yes. The name of the step function execution will be the batch because the name needs to be unique. At the end of our ingest process if there's still items for the group in the lock table, we'll start a new execution of the preingest step function with the number on the end increased.